job-runner/chain/traverser.go

```diff
@@ -81,209 +110,169 @@ func (f *traverserFactory) Make(jobChain proto.JobChain) (Traverser, error) {
-				continue JOB_REAPER
-			}
-
-			// If the failed job is not part of a retryable sequence, then ignore
-			// subsequent jobs. For example, if job B of A -> B -> C  fails, then C is
-			// not ran.
-			jLogger.Warn("job did not complete successfully")
-			continue JOB_REAPER
+	// Suspend running job chain if Job Runner shuts down.
+	go func() {
+		select {
+		case <-t.shutdownChan:
+			t.shutdown()
+		case <-t.doneChan:
+			return
 		}
+	}()
 
-		// Job completed successfully, so enqueue/run the next jobs in the chain.
-		// This will yield multiple next jobs when the current job is the start of
-		// a sequence (a fanout node).
-		jLogger.Infof("job completed successfully")
-
-	NEXT_JOB:
-		for _, nextJ := range t.chain.NextJobs(doneJob.Id) {
-			nextJLogger := jLogger.WithFields(log.Fields{"next_job_id": nextJ.Id})
-
-			// Check to make sure the job is ready to run. It might not be if it has
-			// upstream dependencies that are still running.
-			if !t.chain.JobIsReady(nextJ.Id) {
-				nextJLogger.Infof("next job is not ready to run - not enqueuing it")
-				continue NEXT_JOB
-			}
-
-			nextJLogger.Infof("next job is ready to run - enqueuing it")
-
-			// Copy the jobData from the job that just finished to the next job.
-			// It is important to note that when a job has multiple parent
-			// jobs, it will get its jobData from whichever parent finishes
-			// last. Therefore, a job should never rely on jobData that was
-			// created during an unrelated sequence at any time earlier in
-			// the chain.
-			for k, v := range doneJob.Data {
-				nextJ.Data[k] = v
+	// Add the first jobs to the runJobChan.
+	// Prevent Stop()/shutdown() from closing runJobChan and don't enqueue jobs
+	// if the traverser has already been stopped or shutdown.
+	select {
+	case <-t.stopChan:
+		return
+	case <-t.shutdownChan:
+		return
+	default:
+		for _, job := range jobsToRun {
+			t.logger.Infof("sending initial job (%s) to runJobChan", job.Id)
+			if t.chain.IsSequenceStartJob(job.Id) {
+				// Starting a sequence, so increment sequence try count.
+				t.chain.IncrementSequenceTries(job.Id)
+				seqLogger := t.logger.WithFields(log.Fields{"sequence_id": job.SequenceId, "sequence_try": t.chain.SequenceTries(job.Id)})
+				seqLogger.Info("starting try of sequence")
 			}
-
-			t.runJobChan <- nextJ // add the job to the run queue
+			t.runningJobs[job.Id] = struct{}{}
+			t.runJobChan <- job
 		}
-
-		// Update chain repo for jobs next jobs just started ^
-		t.chainRepo.Set(t.chain)
 	}
 
-	return nil
+	// Start a goroutine to reap done jobs. The Reaper consumes from doneJobChan
+	// and sends the next jobs to be run to runJobChan.
+	t.reaper = t.reaperFactory.MakeRunning(t.runningJobs)
+	go t.reaper.Run()
+
+	<-t.doneChan
 }
 
-// Stop stops the traverser if it's running.
+// Stop stops a running chain by stopping all its running jobs.
 func (t *traverser) Stop() error {
+	// Don't do anything if traverser has already been stopped.
```

Comment:
This isn't thread-safe. Pretend 3 calls to Stop at same time. All three will hit default case on line 259, make a new reaper, and close the chan--2 of them panic on close. Need a mutex and a t.stopped "flag" to serialize access and guarantee only-once execution of this func. (Or the the atomic.Int trick, but haven't found a good example yet.)